sources/engine/Stride.Physics/Elements/CharacterComponent.cs

```diff
@@ -300,6 +300,54 @@ protected override void OnAttach()
             Simulation.AddCharacter(this, (CollisionFilterGroupFlags)CollisionGroup, CanCollideWith);
         }
 
+        /// <summary>
+        /// Reconstruct of character controller to avoid possible UAF issues
+        /// Context: On ColliderShape changes, when ComposeShape is ran to rebuild ColliderShape properties, disposing the old ColliderShape can cause UAF
+        /// issues inside KinematicCharacter and inside Simulation discreteDynamicWorld due to old disposed references to the native object.
+        /// </summary>
+        public override void ComposeShape()
+        {
+            //Disposing of ColliderShape should happen before we remove NativeCollisionObject and KinematicCharacter from Simulation
+            base.ComposeShape();
+
+            //make PairCachingGhostObject references in KinematicCharacter valid, therefore make new instance of kinematic character controller with  
+            //updated NativeCollisionObject keep references valid
+            if (KinematicCharacter != null)
+            {
+                //very mediocre workaround to avoid the nullref when we remove character
+                Simulation simRef = Simulation;
+
+                //remove references in discreteDynamicsWorld
+                Simulation.RemoveCharacter(this);
+                Simulation = simRef;
+
+                //destroy and deref KinematicCharacter then reconstruct it
+                BulletSharp.KinematicCharacterController kinematicCharacterProperties = KinematicCharacter;
+                KinematicCharacter.Dispose();
+                KinematicCharacter = null;
+                BulletSharp.Math.Vector3 unitY = new BulletSharp.Math.Vector3(0f, 1f, 0f);
+                //Make new KinematicCharacter
+                KinematicCharacter = new BulletSharp.KinematicCharacterController((BulletSharp.PairCachingGhostObject)NativeCollisionObject, (BulletSharp.ConvexShape)ColliderShape.InternalShape, StepHeight, ref unitY);
+                OverrideKinematicCharacterValues(kinematicCharacterProperties);
+
+                //now we can add these references BACK in Simulation's discreteDynamicsWorld
+                Simulation.AddCharacter(this, (CollisionFilterGroupFlags)CollisionGroup, CanCollideWith);
+            }
```

Comment:
It seems like virtualizing `ComposeShape` may not be necessary as that method runs right before `OnAttach`, 
https://github.com/stride3d/stride/blob/4ebbf54839cf762ef8e9d135c78f453cf30a723f/sources/engine/Stride.Physics/Engine/PhysicsComponent.cs#L660-L672
So, it you could move this section inside `OnAttach` instead and have it work the same way on top of being able to reduce logic duplication ? I might have missed something though ... let me know !

Reply:
That is a good thing to point out. However due to the fact that users running `ComposeShape` directly in their projects for this scenario can cause the crash, I ended up using the virtual `ComposeShape` approach. Sorry for that delay, I wanted to run some more tests just to be sure and wasnt at my normal computer.

Reply:
Makes sense, good thinking - we're all working on our free time, don't worry about the delay !